### PR TITLE
feat(project_context): repo-identity population + auto-tag + backfill + migrate provenance (#970)

### LINKED ISSUE #970: feat(project_context): define population convention + auto-tag on write + backfill migration (#858 follow-up)
## Summary

`project_context` (added in #858, v3.2) is currently dead weight: the column exists, the retrieval filter reads it, but **nothing ever writes a non-empty value**, and there is no documented convention for what it should hold. This issue proposes to (a) decide the population convention, (b) auto-populate on write, (c) ship a backfill migration for existing stores, and (d) document it. The goal is defense-in-depth scoping, not a fix for a present leak.

## Background (what #858 shipped)

- Schema: `beliefs.project_context TEXT NOT NULL DEFAULT ''` — `src/aelfrice/store.py:619`.
- Resolver: `active_project_context()` reads the `AELFRICE_PROJECT_CONTEXT` env var; unset/blank → `''` → "no filter" — `src/aelfrice/db_paths.py:96`.
- Retrieval filter: `_filter_by_project_context()` keeps a `scope='project'` belief iff `project_context == '' OR project_context == active`; short-circuits to a no-op when the active context is empty — `src/aelfrice/hook.py:1676` (esp. the `if not active: return hits` at :1708).
- Intent per the docstrings: a **within-repo** slice so two worktrees of the *same* repo can see different slices of the *same shared* DB. Cross-**repo** isolation is a separate mechanism: `db_path()` resolves a physically distinct `<git-common-dir>/aelfrice/memory.db` per repo — `src/aelfrice/db_paths.py:57`.

## Observed gap

Across real-world stores, **100% of rows have `project_context = ''`**. Consequences:

1. No writer populates it. New beliefs always land as `''`. The field can only ever be set by a manual `UPDATE` today.
2. No convention. It is undefined whether the field should hold a *repo identity* or a *finer within-repo slice*. These conflict at the data layer (single-string equality match: a row holds one or the other, not both).
3. No defense-in-depth. The only thing keeping repo A's beliefs out of repo B's retrieval is the physical DB separation in `db_path()`. That holds for routine sessions, but **`aelf migrate` deliberately copies beliefs from one DB into another**, collapsing two repos' beliefs into one store where every row is `''` and therefore mutually visible. Federation / peer-scope retrieval is a second path with the same property.

## Design decisions to settle (the crux)

1. **What does `project_context` hold — repo identity or within-repo slice?** If we standardize on repo identity (e.g. the repo's canonical name), it doubles as merge/federation provenance, but it forecloses the originally-intended fine-grained within-repo slicing unless we add a second axis. If we keep it as a within-repo slice, repo identity needs a *different* provenance column for the merge case. Recommend picking one explicitly and documenting it.
2. **Derivation source for repo identity.** Candidates: git remote URL, repo-root basename, or a hash of the git-common-dir (the same identity `db_path()` already computes). Reusing the existing identity function is preferable to inventing a new one.
3. **Locked / user-global beliefs.** L0 user-locked truths are stored as `scope='project'`, so a naive repo-tag backfill would scope global user truths (preferences, profile facts) to a single repo and hide them elsewhere after a merge. Options: leave locked rows `''` (always-visible), or introduce a proper `scope='user'`/global tier. The filter currently only exempts `scope != 'project'` (`hook.py:1712`).
4. **Active-context resolution.** Today env-var only. The #858 docstring already flags a follow-up `.aelfrice/context` state file + `aelf context set <name>` CLI (`db_paths.py:113`). Decide whether population implies a default active context (e.g. resolver defaults to repo identity when the env var is unset) so the field is actually consulted, not just stored.

## Proposed scope

- [ ] Decide and document the convention (decisions 1–4 above) in `docs/`.
- [ ] Auto-populate `project_context` on every belief write (ingest + commit-ingest + promotion paths) from the chosen derivation. Writers: `src/aelfrice/ingest.py`, `src/aelfrice/hook_commit_ingest.py`, `src/aelfrice/store.py` insert path.
- [ ] Backfill migration for existing stores (idempotent; only touches rows where `project_context = ''`). Reversible by design (`SET project_context=''`).
- [ ] Make `aelf migrate` preserve/stamp source provenance into `project_context` (or the new provenance column) so a merge does not collapse to `''`.
- [ ] Tests: writer populates the field; retrieval filter honors it; migrate preserves provenance; locked/global handling matches the decision.

## Non-goals

- Not a fix for an active leak. Routine per-repo isolation via `db_path()` is working as designed.
- Federation transport changes are out of scope beyond ensuring the scope/context check is correct for peer hits.

## References

- #858 — original `project_context` retrieval filter.
- `src/aelfrice/db_paths.py`, `src/aelfrice/hook.py:1676`, `src/aelfrice/store.py:619`, `src/aelfrice/models.py:323`.

### PR DESCRIPTION
**Pull request by robotrocketscience**
Closes #970.

## Problem
`project_context` (added in #858, v3.2) was dead weight: the column existed and the retrieval filter read it, but **nothing ever wrote a non-empty value** (100% of rows `''`), and `aelf migrate` dropped the column, collapsing two repos' beliefs into one mutually-visible pool — the one case physical DB separation can't cover.

## Convention chosen (ADR 0003)
The four #970 "crux" decisions, resolved per the issue's recommendations:

1. **What it holds → repo identity** (not a within-repo slice). Format `<repo-root-basename>-<8 hex blake2b of the abs git-common-dir>`.
2. **Derivation → reuse the git-common-dir** `db_path()` already keys on (no new identity source). Two worktrees of one repo share an identity.
3. **Locked/global rows → stay `''`** (always-visible). A repo-tag backfill must not scope a user's global truths to one repo where `aelf migrate` would then hide them.
4. **Active-context resolution → opt-in.** The resolver default stays env-driven (`unset = no filter`); scoping activates by exporting `AELFRICE_PROJECT_CONTEXT="$(repo identity)"`.

## What this PR does (4 atomic commits)
- **feat(store):** `insert_belief` auto-stamps the store's repo identity onto eligible new beliefs; an idempotent, reversible, locked-exempt backfill stamps pre-existing rows on first open. No-op when no identity is set (direct opens) → exact pre-#970 behaviour preserved.
- **feat(db_paths):** `repo_identity()` / `repo_identity_from_db_path()` (the latter derives identity from the store's on-disk layout with **no extra git fork**); `_open_store()` injects it.
- **feat(migrate):** `aelf migrate` now preserves an already-stamped `project_context` and stamps source-`''` eligible rows with the **source** repo's identity, so provenance survives the merge instead of collapsing to `''`.
- **docs(adr):** ADR 0003 records the convention; Belief/resolver docstrings reconciled.

## Decision 4 — flagged for your call
I implemented decision 4 as **opt-in** (column populated + migrate-safe, but the resolver default left env-driven). I deliberately did **not** flip the resolver to default to repo identity, because that silently reverses #858's documented "unset = no filter" contract and changes retrieval for every store (a no-op for a single repo, but a semantics change). If you want auto-consult-by-default, that's a one-line resolver change + filter-test update — say the word and I'll add a commit.

## Acceptance (#970 proposed scope)
- [x] Convention decided and documented (ADR 0003).
- [x] Auto-populate `project_context` on write (store choke point covers ingest/commit-ingest/promotion).
- [x] Idempotent, reversible backfill for existing stores.
- [x] `aelf migrate` preserves/stamps source provenance.
- [x] Tests: writer populates; backfill eligibility (locked/global exempt); migrate preserves provenance.

## Verification
- New: `test_store_project_context_970.py` (9), `test_db_paths_repo_identity_970.py` (9), `test_migrate_project_context_970.py` (4).
- Regression sweep across store/ingest/migrate/federation/cli/hook suites: green (full-suite result in PR gate).

## Summary by Sourcery

Populate and preserve the `project_context` column using a stable repo identity so beliefs gain per-repo provenance without changing default retrieval behaviour.

New Features:
- Auto-stamp new project-scope, non-user-locked beliefs with a repo identity–based project_context derived from the store’s on-disk git layout.
- Expose helpers to compute a stable repo identity from the git common directory or an already-resolved DB path and integrate this identity when opening stores.
- Have `aelf migrate` carry forward existing project_context values and tag previously empty project-scope beliefs with the source store’s repo identity to retain provenance across merges.
- Document the chosen convention for project_context as repo identity in ADR 0003 and update Belief documentation accordingly.

Enhancements:
- Add an idempotent, reversible backfill to stamp repo identity into eligible existing beliefs while leaving locked and global rows always-visible.
- Keep project-context scoping opt-in by retaining the environment-driven active_project_context resolver while ensuring the column is consistently populated.

Tests:
- Add tests validating store-side project_context stamping and backfill behaviour for various scopes and lock levels.
- Add tests covering repo-identity derivation from git paths and DB paths, and verifying that store opens in repo layout propagate the identity into writes.
- Add tests ensuring migrate preserves or stamps project_context correctly for stamped, empty, locked, global, and non-repo source rows.

## Summary by CodeRabbit

* **New Features**
  * Beliefs are now automatically tagged with repository identity information for improved tracking and scoping
  * Migration preserves or derives repository identity for eligible beliefs

* **Documentation**
  * Added Architecture Decision Record (ADR 0003) documenting project context and repo identity conventions
  * Updated field documentation with new repository identity stamping behavior

* **Tests**
  * Comprehensive coverage for repository identity derivation, store stamping, and migration behavior

### CODE REVIEW

**Comment by robotrocketscience** [comment]
[claim:review:Kulili:2026-06-18T23:42:23Z]

**Comment by robotrocketscience** [comment]
Reviewed the diff against `github/main`. Approving.

**Correctness checks that pass:**
- **Identity consistency.** `repo_identity()` (git-fork) and `repo_identity_from_db_path()` (path-derived) both BLAKE2b-digest `str(<resolved git-common-dir>)`. On the normal path `db_path()` returns `<git_dir>/aelfrice/memory.db`, so `p.parent.parent` reconstructs the same `git_dir` string the fork path resolves — the opt-in filter will match the stamped value. (With an `AELFRICE_DB` override outside the `aelfrice/` layout the stamp is `''`, which keeps rows always-visible — a safe no-op, not silent hiding.)
- **Migrate preserves provenance, no double-stamp.** The target is opened bare (`MemoryStore(str(target_path))`, no `project_context_default`), so `_project_context_for_insert` honours the source-derived value verbatim and never re-stamps with the destination identity. Source-`''` locked/global and non-repo-source rows correctly stay `''`.
- **Locked/global exemption is consistent** between the writer (`_project_context_for_insert`), the backfill UPDATE (`scope = project AND lock_level != user`), and migrate's read path.
- **Backfill is idempotent + reversible + no-op on direct opens** (empty default short-circuits). Pre-#970 behaviour for non-repo stores is exactly preserved; retrieval is unchanged until `AELFRICE_PROJECT_CONTEXT` is exported.

**Verification:** CI green (pytest 3.12/3.13, migration-policy-check, full staging gate). 3 new test files cover writer/backfill, identity derivation, and migrate provenance. Discretion grep on the diff: clean.

Decision 4 (resolver default left env-driven rather than flipped to auto-consult) is the right call to defer to the operator — flipping it would silently reverse #858's "unset = no filter" contract. Not a blocker.

File: src/aelfrice/db_paths.py [hunk 1]
```diff
@@ -78,11 +83,62 @@ def _ensure_parent_dir(path: Path) -> None:
...
+    Format: ``<repo-root-basename>-<8 hex>``. git-common-dir is
+    ``<root>/.git`` (shared across a repo's worktrees), so its parent's
+    name is the repo root basename — included for human legibility in
+    `aelf` output and migrate provenance. The 8-hex BLAKE2b digest of the
+    absolute git-common-dir disambiguates two same-named repos.
+    """
+    root = git_dir.parent
+    basename = root.name or "repo"
+    digest = hashlib.blake2b(str(git_dir).encode("utf-8"), digest_size=4).hexdigest()
+    return f"{basename}-{digest}"
```
[the remark this answers is not in the document]
Comment (robotrocketscience) [reply src/aelfrice/db_paths.py:106]: Good catch, valid bug_risk: `migrate --from` passes an unresolved `Path(args.from_path)`, so a relative legacy path would digest differently than `repo_identity()`. Fixed in 83528485 by resolving `git_dir` to absolute before hashing (idempotent for the already-resolved callers).

File: docs/adr/0003-project-context-repo-identity-convention.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+Mechanics: `insert_belief` stamps the store's repo identity onto new eligible rows; a one-shot idempotent, reversible backfill stamps pre-existing eligible rows on first open; `aelf migrate` preserves an already-stamped value and stamps source-`''` eligible rows with the *source* repo's identity so provenance survives the merge.
+
+## Alternatives considered
+
+- **Within-repo slice (keep #858 intent).** Rejected: nothing populated it for nine months, and it provides no merge/federation provenance, which was the concrete gap. An explicit `project_context` value on a belief is still honoured verbatim, so per-slice tagging remains possible by opting in; it is just not the default.
+- **Flip the resolver default to repo identity (auto-consult, decision 4 = yes).** Rejected for now: it reverses the documented "unset = no filter" contract and changes retrieval behaviour for every store (a no-op for a single repo, but a silent semantics change). Left as a follow-up the operator can ratify; the column and migrate provenance land first.
```
[the remark this answers is not in the document]
Comment (robotrocketscience) [reply docs/adr/0003-project-context-repo-identity-convention.md]: Applied in 198baee9.
